File: src/test/starterkit/starter.cpp
#include <iostream>
#include <string>
#include <vector>
#include <list>
#include <algorithm>

using namespace std;

enum BuildingType {
    HQ
};

enum CommandType {
    WAIT,
    MOVE,
    TRAIN
};

ostream &operator<<(ostream &os, CommandType cmdType) {
    switch (cmdType) {
        case WAIT:
            return os << "WAIT";
        case MOVE:
            return os << "MOVE";
        case TRAIN:
            return os << "TRAIN";
    }
    return os;
}

class Position {
public:

    int x, y;

    Position(int x, int y) : x(x), y(y) {
    }
};

class Command {

public:

    CommandType t;
    Position p;
    int idOrLevel;

    Command(CommandType t, int idOrLevel, const Position &p) : t(t), idOrLevel(idOrLevel), p(p) {
    }

    void print() {
        cout << t << " " << idOrLevel << " " << p.x << " " << p.y << ";";
    }
};

class Unit {

public:

    int id;
    int owner;
    int level;
    Position p;

    Unit(int x, int y, int id, int level, int owner) : p(x, y), id(id), level(level), owner(owner) {
    }

    void debug() {
        cerr << "unit of level " << level << " at " << p.x << " " << p.y << " owned by " << owner;
    }

    bool isOwned() {
        return owner == 0;
    }
};

class Building {

public:

    Position p;
    BuildingType t;
    int owner;

    Building(int x, int y, int t, int owner) : p(x, y), t(static_cast<BuildingType>(t)), owner(owner) {
    }

    void debug() {
        cerr << t << " at " << p.x << " " << p.y << " owned by " << owner;
    }

    bool isHQ() {
        return t == HQ;
    }

    bool isOwned() {
        return owner == 0;
    }
    
};

class Game {

    list<Unit> units;

    list<Building> buildings;

    int gold, income;

    list<Command> commandes;

public:

    void debug() {
        for_each(units.begin(), units.end(), [](Unit &u) { u.debug(); });
        for_each(buildings.begin(), buildings.end(), [](Building &u) { u.debug(); });
    }

    // not useful in Wood 3
    void init() {
        int numberMineSpots;
        cin >> numberMineSpots; cin.ignore();
        for (int i = 0; i < numberMineSpots; i++) {
            int x;
            int y;
            cin >> x >> y; cin.ignore();
        }
    }

    void update() {

        units.clear();
        buildings.clear();
        commandes.clear();

        // READ TURN INPUT
        cin >> gold; cin.ignore();
        cin >> income; cin.ignore();

        int opponentGold;
        cin >> opponentGold; cin.ignore();
        int opponentIncome;
        cin >> opponentIncome; cin.ignore();

        for (int i = 0; i < 12; i++) {
            string line;
            cin >> line; cin.ignore();
            cerr << line << endl;
        }
        int buildingCount;
        cin >> buildingCount; cin.ignore();
        for (int i = 0; i < buildingCount; i++) {
            int owner;
            int buildingType;
            int x;
            int y;
            cin >> owner >> buildingType >> x >> y; cin.ignore();
            buildings.push_back(Building(x, y, buildingType, owner));
        }
        int unitCount;
        cin >> unitCount; cin.ignore();
        for (int i = 0; i < unitCount; i++) {
            int owner;
            int unitId;
            int level;
            int x;
            int y;
            cin >> owner >> unitId >> level >> x >> y; cin.ignore();
            units.push_back(Unit(x, y, unitId, level, owner));
        }
    }

    void buildOutput() {
        // @TODO "core" of the AI
        trainUnits();
        
        moveUnits();
    }

    void output() {
        for_each(commandes.begin(), commandes.end(), [](Command &c) {
            c.print();
        });
        cout << "WAIT" <<  endl;
    }


private:

    void trainUnits() {
        Position trainingPosition = findTrainingPosition();

        if (gold > 30) {
            commandes.push_back(Command(TRAIN, 1, trainingPosition));
        }
    }
    
    // move to the center
    void moveUnits() {
        Position center(5, 5);

        for (auto &unit : units) {
            if (unit.isOwned()) {
                commandes.push_back(Command(MOVE, unit.id, center));
            }
        }
    }

    // train near the HQ for now
    Position findTrainingPosition() {
        const Building &HQ = getHQ();

        if (HQ.p.x == 0) {
            return Position(0, 1);
        }
        return Position(11, 10);
    }

    const Building &getHQ() {
        for (auto &b : buildings) {
            if (b.isHQ() && b.isOwned()) {
                return b;
            }
        }
    }

    const Building &getOpponentHQ() {
        for (auto &b : buildings) {
            if (b.isHQ() && !b.isOwned()) {
                return b;
            }
        }
    }
};

int main()
{
    Game g;
    g.init();

    // game loop
    while (true) {
        g.update();
        //g.debug();
        g.buildOutput();
        g.output();
    }
    return 0;
}

File: src/test/starterkit/starter.java
import java.util.*;
import java.io.*;
import java.math.*;

class Player {

    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        
        Game g = new Game();
        g.init(in);
        
        // game loop
        while (true) {
            
            g.update(in);
            //g.debug();
            g.buildOutput();
            g.output();
        }
    }
}

class Game {
    
    List<Unit> units;
    
    List<Building> buildings;
    
    int gold, income;
    
    List<Command> output;
    
    public Game() {
        units = new ArrayList<Unit>();
        buildings = new ArrayList<Building>();
        output = new ArrayList<Command>();
    }
    
    // not useful in Wood 3
    public void init(Scanner in) {
        int numberMineSpots = in.nextInt();
        for (int i = 0; i < numberMineSpots; i++) {
            int x = in.nextInt();
            int y = in.nextInt();
        }
    }
    
    public void update(Scanner in) {

        units.clear();
        buildings.clear();
        output.clear();
        
        // READ TURN INPUT
        gold = in.nextInt();
        int income = in.nextInt();
        int opponentGold = in.nextInt();
        int opponentIncome = in.nextInt();
        for (int i = 0; i < 12; i++) {
            String line = in.next();
            System.err.println(line);
        }
        int buildingCount = in.nextInt();
        for (int i = 0; i < buildingCount; i++) {
            int owner = in.nextInt();
            int buildingType = in.nextInt();
            int x = in.nextInt();
            int y = in.nextInt();
            buildings.add(new Building(x, y, buildingType, owner));
        }
        int unitCount = in.nextInt();
        for (int i = 0; i < unitCount; i++) {
            int owner = in.nextInt();
            int unitId = in.nextInt();
            int level = in.nextInt();
            int x = in.nextInt();
            int y = in.nextInt();
            units.add(new Unit(x, y, unitId, level, owner));
        }
    }
    
    public void buildOutput() {
        // @TODO "core" of the AI
        trainUnits();
        
        moveUnits();
    }
    
    private void trainUnits() {
        Position trainingPosition = findTrainingPosition();
        if (gold > 30) {
            output.add(new Command(CommandType.TRAIN, 1, trainingPosition));
        }
    }
    
    // move to the center
    private void moveUnits() {
        Position center = new Position(5,5);
        units.stream()
            .filter(
                u -> u.isOwned())
            .forEach(
                myUnit -> output.add(new Command(CommandType.MOVE, myUnit.id, center))
            );
    }
    
    // train near the HQ for now
    private Position findTrainingPosition() {
        Building HQ = getHQ();
        if (HQ.p.x == 0) {
            return new Position(0, 1);
        }
        return new Position(11, 10);
    }
    
    public void output() {
        StringBuilder s = new StringBuilder("WAIT;");
        output.stream().forEach(c -> s.append(c.get()));
        System.out.println(s);
    }

    public void debug() {
        units.stream().forEach(u -> u.debug());
        buildings.stream().forEach(b -> b.debug());
    }
    
    private Building getHQ() {
        return buildings.stream().filter(b -> b.isHQ() && b.isOwned()).findFirst().get();
    }
    
    private Building getOpponentHQ() {
        return buildings.stream().filter(b -> b.isHQ() && !b.isOwned()).findFirst().get();
    }
}

class Command {
    
    CommandType t;
    
    Position p;
    
    int idOrLevel;
    
    public Command(CommandType t, int idOrLevel, Position p) {
        this.t = t;
        this.p = p;
        this.idOrLevel = idOrLevel;
    }
    
    public String get() {
        return t.toString()+" "+idOrLevel+" "+p.x+" "+p.y+";";
    }
}

class Position {
    
    int x, y;
    
    public Position(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

class Unit {
    
    Position p;
    
    int level;
    
    int owner;
    
    int id;
    
    public Unit(int x, int y, int id, int level, int owner) {
        this.p = new Position(x,y);
        this.id = id;
        this.level = level;
        this.owner = owner;
    }
    
    public void debug() {
        System.err.println("unit of level "+level+" at "+p.x+" "+p.y+" owned by "+owner);
    }
    
    public boolean isOwned() {
        return owner == 0;
    }
}

class Building {
    
    Position p;
    
    BuildingType t;
    
    int owner;
    
    public Building(int x, int y, int t, int owner) {
        this.p = new Position(x,y);
        this.t = BuildingType.get(t);
        this.owner = owner;
    }
    
    public void debug() {
        System.err.println(t+" at "+p.x+" "+p.y+" owned by "+owner);
    }
    
    public boolean isHQ() {
        return t.equals(BuildingType.HQ);
    }
    
    public boolean isOwned() {
        return owner == 0;
    }
    
}

enum BuildingType {
    
    HQ;
    
    static public BuildingType get(int type) {
        switch (type) {
            case 0: 
                return HQ;
        }
        return null;
    }
}

enum CommandType {
    
    MOVE,
    TRAIN;
}